File: stronghold/code/python/ORF.py
import os
import json


class ORFSolution:
    def __init__(self):
        self.dna = None
        self.dna_reverse_comple = None
        self.proteins = set()
        self.codons = {}

        self.get_dna_sequence()
        self.get_codons()
        self.open_reading_frame()

    def get_complement_of(self, dna: str):
        complement = {'A': 'T', 'T': 'A', 'G': 'C', 'C': 'G'}

        c = ""
        for base in dna:
            c += complement[base]

        # retornando o complemento
        return c

    # complemento reverso
    def reverse(self, complement: str):
        # tamanho do complemento
        comple_size = len(complement)

        aux = [None] * comple_size

        for i in range(comple_size//2+1):
            j = comple_size - i - 1
            # trocando as posições
            aux[j], aux[i] = complement[i], complement[j]

        return ''.join(aux)

    # lendo a tabela de códons
    def get_codons(self):
        try:
            path = os.path.realpath("../../../utils/codons.json")
            input_data = open(path, 'r')
            self.codons = json.load(input_data)
            input_data.close()
        except IOError:
            print("Erro na leitura do arquivo .json!")

    # recebendo a sequência de DNA
    def get_dna_sequence(self):
        try:
            path = os.path.realpath("../../input/rosalind_orf.txt")
            input_data = open(path, 'r')
            # linhas de texto do arquivo
            lines = [line.strip() for line in input_data.readlines()]
            # sequência de DNA
            dna = ''.join(lines[1::])
            self.dna = dna
            # complemento
            complement = self.get_complement_of(dna)
            # complemento reverso
            self.dna_reverse_comple = self.reverse(complement)
            input_data.close()
        except IOError:
            print("Erro na leitura do arquivo!")

    # lendo os frames
    def reading_frames(self, sequence):
        start_codon = "ATG"
        stop_codon = ["TAG", "TGA", "TAA"]
        begins = []
        ends = []

        for start in range(3):
            # tamanho da sequência
            seq_size = len(sequence[start::])
            end = seq_size - seq_size % 3
            orf_starts = []

            for i in range(start, end, 3):
                codon = sequence[i:i+3]

                if codon == start_codon:
                    # adicionando as posições iniciais de cada ORF
                    orf_starts.append(i)

                if codon in stop_codon:
                    # verificando se alguma posição inicial foi adicionada
                    if len(orf_starts) != 0:
                        begins.append(orf_starts)
                        ends.append(i)
                    # limpando as posições iniciais
                    orf_starts = []

        positions = {"begins": begins, "ends": ends}

        return positions

    # DNA -> RNA

    def transcribe_dna(self, dna: str):
        return dna.replace('T', 'U')

    # RNA -> PROTEÍNA
    def translate_rna(self, rna: str):
        protein = ""
        for i in range(0, len(rna), 3):
            codon = rna[i:i+3]
            protein += self.codons[codon]
        return protein

    def get_orf(self, sequence: str, positions: dict):
        begins = positions['begins']
        ends = positions['ends']

        for begin, end in zip(begins, ends):
            for i in begin:
                dna = sequence[i:end]
                rna = self.transcribe_dna(dna)
                protein = self.translate_rna(rna)
                self.proteins.add(protein)

    def open_reading_frame(self):
        # dna
        positions_one = self.reading_frames(self.dna)
        # complemento reverso do DNA
        positions_two = self.reading_frames(self.dna_reverse_comple)

        self.get_orf(self.dna, positions_one)
        self.get_orf(self.dna_reverse_comple, positions_two)

    def get_protein(self):
        try:
            path = os.path.realpath("../../output/ORF.txt")
            output_data = open(path, 'w')
            for protein in self.proteins:
                output_data.write(f"{protein}\n")
            output_data.close()
        except IOError:
            print("Erro na escrita do arquivo!")


if __name__ == "__main__":
    solution = ORFSolution()
    solution.get_protein()

File: stronghold/code/c++/ORF.cpp
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <algorithm>
#include <map>

class ORFSolution
{
private:
    std::string dna_seq, dna_reverse_comp_seq, start_codon = "ATG";
    // proteínas
    std::vector<std::string> proteins;
    // não codificam para nenhum aminoácido
    std::vector<std::string> stop_codons = {"TAA", "TAG", "TGA"};
    std::map<char, char> complements = {
        {'A', 'T'}, {'T', 'A'}, {'G', 'C'}, {'C', 'G'}};

    // codificam para algum aminoácido
    std::map<std::string, char> rna_codons = {
        {"UUU", 'F'}, {"CUU", 'L'}, {"AUU", 'I'}, {"GUU", 'V'}, {"UUC", 'F'}, {"CUC", 'L'}, 
        {"AUC", 'I'}, {"GUC", 'V'}, {"UUA", 'L'}, {"CUA", 'L'}, {"AUA", 'I'}, {"GUA", 'V'}, 
        {"UUG", 'L'}, {"CUG", 'L'}, {"AUG", 'M'}, {"GUG", 'V'}, {"UCU", 'S'}, {"CCU", 'P'}, {"ACU", 'T'}, {"GCU", 'A'}, {"UCC", 'S'}, {"CCC", 'P'}, {"ACC", 'T'}, {"GCC", 'A'}, {"UCA", 'S'}, {"CCA", 'P'}, {"ACA", 'T'}, {"GCA", 'A'}, {"UCG", 'S'}, {"CCG", 'P'}, {"ACG", 'T'}, {"GCG", 'A'}, {"UAU", 'Y'}, {"CAU", 'H'}, {"AAU", 'N'}, {"GAU", 'D'}, {"UAC", 'Y'}, {"CAC", 'H'}, {"AAC", 'N'}, {"GAC", 'D'}, {"CAA", 'Q'}, {"AAA", 'K'}, {"GAA", 'E'}, {"CAG", 'Q'}, {"AAG", 'K'}, {"GAG", 'E'}, {"UGU", 'C'}, {"CGU", 'R'}, {"AGU", 'S'}, {"GGU", 'G'}, {"UGC", 'C'}, {"CGC", 'R'}, {"AGC", 'S'}, {"GGC", 'G'}, {"CGA", 'R'}, {"AGA", 'R'}, {"GGA", 'G'}, {"UGG", 'W'}, {"CGG", 'R'}, {"AGG", 'R'}, {"GGG", 'G'}};

    std::vector<std::string>::iterator it;

public:
    ORFSolution()
    {
        read_dna_seq();
        open_reading_frame();
    }

    // inverter sequência
    void reverse(std::string &complement)
    {
        int n = complement.length();

        for (int i = 0; i < n / 2; ++i)
        {
            std::swap(complement[i], complement[n - i - 1]);
        }
    }

    // definir complemento
    std::string get_complement(std::string dna)
    {
        std::string dna_comp;

        for (auto base : dna)
        {
            dna_comp += complements[base];
        }

        return dna_comp;
    }

    // transcrever sequência de DNA
    void transcribe_dna(std::string &dna)
    {
        // T(timina) -> U(uracila)
        std::replace(dna.begin(), dna.end(), 'T', 'U');
    }

    // traduzir RNA
    void translate_rna(std::string &rna)
    {
        std::string protein;
        for (unsigned i = 3; i < rna.length(); i += 3)
        {
            // códon da sequência
            std::string codon = rna.substr(i - 3, 3);

            protein += rna_codons[codon];
        }

        if (!protein.empty())
        {
            proteins.push_back(protein);
        }
    }

    // lendo a sequência de DNA
    void read_dna_seq()
    {
        std::ifstream input_data("../../input/rosalind_orf.txt");

        std::string line;

        if (input_data.is_open())
        {
            // lendo as linhas do arquivo
            while (std::getline(input_data, line, '\n'))
            {
                // verificando se não é o rótulo
                if (line[0] != '>')
                {
                    // lendo a sequência do DNA
                    dna_seq += line;
                }
            }
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo!" << std::endl;
        }

        // definindo o complemento
        std::string dna_comp = get_complement(dna_seq);
        // invertendo a sequência
        reverse(dna_comp);
        // atualizando a variável
        dna_reverse_comp_seq = dna_comp;

        std::cout << dna_seq << std::endl;
        std::cout << dna_reverse_comp_seq << std::endl;
    }

    // lendo os frames
    void read_frames(std::string dna, unsigned start)
    {
        std::string rna;
        int init = -1;
        for (unsigned i = start + 3; i < dna.length(); i += 3)
        {
            std::string codon = dna.substr(i - 3, 3);

            // começo da sequência
            if (codon == start_codon)
            {
                init = 1;
            }

            // verificando se o códon da sequência não codifica para algum aminoácido
            it = std::find(stop_codons.begin(), stop_codons.end(), codon);

            // término da sequência
            if (it != stop_codons.end())
            {
                init = 0;
                std::cout << "RNA: " << rna << std::endl;


                // processo de tradução
                translate_rna(rna);
                // limpando a sequência
                rna.clear();
            }

            if (init == 1)
            {
                rna += codon;
            }
        }

        // "TAA", "TAG", "TGA"
        // ATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA        ATGATCCGAGTAGCATCT
        // ATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAA GCCTGA ATGATCCGAGTAGCATCTCAG
    }

    void open_reading_frame()
    {
        // frames
        read_frames(dna_seq, 0);
        read_frames(dna_seq, 1);
        read_frames(dna_seq, 2);

        read_frames(dna_reverse_comp_seq, 0);
        read_frames(dna_reverse_comp_seq, 1);
        read_frames(dna_reverse_comp_seq, 2);
    }

    void get_proteins()
    {
        std::ofstream output_data("../../output/ORF.txt");

        if (output_data.is_open())
        {
            for (auto protein : proteins)
            {
                output_data << protein << '\n';
            }
        }
        else
        {
            std::cerr << "Não foi possível abrir o arquivo!" << std::endl;
        }
    }
};

int main()
{
    ORFSolution solution;
    solution.get_proteins();

    return 0;
}